### PR TITLE
[BREAKING] Enable access to service internals from Auth funcs

### PR DESCRIPTION
**Pull request by difengo**
Current implementation of Auth funcs doesn't allow access to the service logger or to any other internal property.

This update: 
- renames and turns the Auth funcs (auth.go) to service methods 
- simplifies the endpoints creation statement in the server example main.go file

This is accomplished by using a new separate "Auther" interface  in the generated service.go file that implements the Auth functions.

Then, considering the security example (goa/examples/security), the signature of the Auth functions (goa/examples/security/auth.go) are then changed:

From:
```go
// SecuredServiceBasicAuth implements the authorization logic for service
// "secured_service" for the "basic" security scheme.
func SecuredServiceBasicAuth(ctx context.Context, user, pass string, s *security.BasicScheme) (context.Context, error) {
...
}
```

To:
```go
// BasicAuth implements the authorization logic for service "secured_service"
// for the "basic" security scheme.
func (s *securedServiceSvc) BasicAuth(ctx context.Context, user, pass string, scheme *security.BasicScheme) (context.Context, error) {
...
}
```

In the generated main.go file (goa/examples/security/cmd/multiauth/main.go) passing the reference of the Auth funcs is no more necessary for the endpoints creation and changed:

From:
```go
// Wrap the services in endpoints that can be invoked from other
// services potentially running in different processes.
var (
	securedServiceEndpoints *securedservice.Endpoints
)
{
	securedServiceEndpoints = securedservice.NewEndpoints(securedServiceSvc, multiauth.SecuredServiceBasicAuth, multiauth.SecuredServiceJWTAuth, multiauth.SecuredServiceAPIKeyAuth, multiauth.SecuredServiceOAuth2Auth)
}
```

To:
```go
// Wrap the services in endpoints that can be invoked from other
// services potentially running in different processes.
var (
	securedServiceEndpoints *securedservice.Endpoints
)
{
	securedServiceEndpoints = securedservice.NewEndpoints(securedServiceSvc)
}
```

### CODE REVIEW

File: codegen/service/service.go [hunk 1]
```diff
@@ -205,6 +206,15 @@ type Service interface {
 {{- end }}
 }
 
+{{- if .Schemes }}
+// authenticated is the internal interface for security functions
```
Comment (raphael) [inline codegen/service/service.go]:
```suggestion
// auther defines the authorization functions used by the service.
```

File: codegen/service/service.go [hunk 2]
```diff
@@ -205,6 +206,15 @@ type Service interface {
...
+type authenticated interface {
```
Comment (raphael) [inline codegen/service/service.go]: `auther` might be a bit more idiomatic?
```suggestion
type auther interface {
```

File: codegen/service/endpoint.go [hunk 3]
```diff
@@ -184,10 +184,14 @@ type {{ .VarName }} struct {
 
 // input: EndpointsData
 const serviceEndpointsInitT = `{{ printf "New%s wraps the methods of the %q service with endpoints." .VarName .Name | comment }}
-func New{{ .VarName }}(s {{ .ServiceVarName }}{{ range .Schemes }}, auth{{ . }}Fn security.Auth{{ . }}Func{{ end }}) *{{ .VarName }} {
+func New{{ .VarName }}(s {{ .ServiceVarName }}) *{{ .VarName }} {
+{{- if .Schemes }}
+	// Typing service to authenticated interface
+	a := s.(authenticated)
```
Comment (raphael) [inline codegen/service/endpoint.go]: As per the comment below:
```suggestion
	a := s.(auther)
```

**Comment by raphael** [comment]
This seems like a good idea. The only thing that makes me a bit uncomfortable is the user experience: how do I know that my service struct needs to implement these functions? Getting to know that though a compilation error isn't great. Would it be possible to add a generated comment somewhere that indicates what needs to be done?

**Comment by raphael** [comment]
No security consequences, at the end of the day user code is implementing them anyway. Thank you!

File: codegen/service/service.go [hunk 4]
```diff
@@ -207,9 +207,10 @@ type Service interface {
 }
 
 {{- if .Schemes }}
-// auther defines the authorization functions used by the service.
-type auther interface {
+// Auther defines the authorization functions to be implemented by the service.
+type Auther interface {
 	{{- range .Schemes }}
+	{{ printf "%sAuth should implement the authorization logic for the %s security scheme." .Type .Type | comment }}
```
Comment (raphael) [inline codegen/service/service.go]: nitpick: could the comment be `%sAuth implements the authorization logic for the %s security scheme."`

Comment (difengo) [reply codegen/service/service.go]: I changed the comment to your suggestion.

By the way, is there a place I could discuss some improvements with you before doing pull requests ? You mentioned a Slack but I didn't find the URL.

Comment (raphael) [reply codegen/service/service.go]: Great thank you! The details on how to join the slack channel are on https://goa.design (search for "slack".  There's also a badge in the README).

**Comment by raphael** [comment]
Sorry one last thing: this is a breaking change in at least a couple of places (`NewEndpoints` signature and `Service` interface). Can you please document the breaking changes and how to upgrade in the PR description? Thank you!

**Comment by difengo** [comment]
PR description updated with examples to better explain the changes.

**Comment by raphael** [comment]
Thank you!

**Pull request merged**